tools/wake/timeline.cpp

```diff
@@ -0,0 +1,242 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <iostream>
+#include <algorithm>
+#include <map>
+#include <fstream>
+
+#include "timeline.h"
+#include "runtime/database.h"
+#include "util/execpath.h"
+
+struct JobNode {
+    JobReflection &job;
+    long parent_id = -1;
+    std::vector<long> dependencies;
+
+    long long starttime;
+    long long endtime;
+
+    long long hypothetical_starttime = 0;
+    long long hypothetical_endtime = -1;
+
+    bool top_sort_visited = false;
+    bool is_on_critical_path = false;
+    bool transitive_reduction_visited = false;
+
+    explicit JobNode(JobReflection &_job) : job(_job) {
+        std::string start_str = std::to_string(_job.starttime.asInt64());
+        std::string end_str = std::to_string(_job.endtime.asInt64());
+        starttime = std::stoll(start_str.substr(0, start_str.length() - 6));
+        endtime = std::stoll(end_str.substr(0, end_str.length() - 6));
+    }
+};
+
+void dfs_top_sort(long job_id, std::map<long, JobNode> &job_map, std::vector<long> &top_sorted_jobs) {
+    JobNode &job = job_map.at(job_id);
+    job_map.at(job_id).top_sort_visited = true;
+    for (auto it = job.dependencies.begin(); it < job.dependencies.end(); it++) {
+        if (job_map.find(*it) == job_map.end()) {
+            job.dependencies.erase(it--);
+            continue;
+        }
+        if (!(job_map.at(*it).top_sort_visited)) {
+            dfs_top_sort(*it, job_map, top_sorted_jobs);
+        }
+    }
+    top_sorted_jobs.emplace_back(job_id);
+}
+
+void top_sort(std::map<long, JobNode> &job_map, std::vector<long> &top_sorted_jobs) {
+    for (const auto &job_pair: job_map) {
+        if (!(job_pair.second.top_sort_visited)) {
+            dfs_top_sort(job_pair.first, job_map, top_sorted_jobs);
+        }
+    }
+}
+
+long fill_one_dependency(std::pair<int, long> access, long dependency, std::map<long, JobNode> &job_map) {
+    if (access.first == 2) {
+        return access.second;
+    }
+    if (job_map.find(access.second) != job_map.end()) {
+        job_map.at(access.second).dependencies.emplace_back(dependency);
+    }
+    return dependency;
+}
+
+void fill_all_dependencies(const std::vector<std::pair<int, long>> &accesses, std::map<long, JobNode> &job_map) {
+    long dependency = -1;
+    for (auto access: accesses) {
+        dependency = fill_one_dependency(access, dependency, job_map);
+    }
+}
+
+void assign_parents(std::map<long, JobNode> &job_map, std::vector<long> &top_sorted_jobs) {
+    for (long job_id: top_sorted_jobs) {
+        JobNode &job_node = job_map.at(job_id);
+        for (long dependency_id: job_node.dependencies) {
+            JobNode dependency = job_map.at(dependency_id);
+            if (dependency.hypothetical_endtime >= job_node.hypothetical_starttime) {
+                job_node.hypothetical_starttime = dependency.hypothetical_endtime;
```

Comment:
I'm confused here, why doe we set the start time to the end time if the end time is greater than the start time?

Reply:
Added a clarifying comment